### PR TITLE
[Android] Register Brave as Android AutofillService for address autofill

### LINKED ISSUE #41334: Re enable autofill service on Android
Parent issue https://github.com/brave/brave-browser/issues/36227

## Description
We had that feature implemented https://github.com/brave/brave-browser/issues/36226, but we had to revert it as users reported crashes in it https://github.com/brave/brave-browser/issues/41254.
I've tried to use it on my Pixel 8 Android 14 and it didn't work, but eventually it started to work for some reason in Chrome.
Android 11 emulator just crashed with the same stack as the user experienced.

From the investigation I did we need to move the `BraveAutofillService` in the base module that installs on the very beginning and that's how other services in Brave and upstream are done.

A PR that removed the autofill feature is https://github.com/brave/brave-core/pull/25759

### PR DESCRIPTION
**Pull request by SergeyZhukovsky**
Re-enables the Android AutofillService that was removed in https://github.com/brave/brave-core/pull/25759 due to crashes on split-APK installs. A sec/privacy review was done here https://github.com/brave/reviews/issues/1520. The current implementation not only re-enables the autofill service but also improves field detection and fill/save logic:
Field detection:
  - Added Layer 1: maps standard Android `AUTOFILL_HINT_*` constants and `W3C` autocomplete tokens (`given-name`, `street-address`, `address-level2`, etc.) to canonical hint keys. Previously, fields with standard hints were not recognized because the old implementation only did keyword substring matching.
  - Added Layer 3: reads Chromium's HTML attributes (`computed-autofill-hints`, `name`, `label`) from WebView nodes — this handles web forms that lack autocomplete attributes, since Chromium still classifies fields internally using ML and heuristics. Previously, such forms produced no autofill suggestions at all.
  - Added duplicate-skipping across all detection layers — when a hint is already taken by another field, deeper layers are tried. This handles forms with incorrect autocomplete attributes (e.g. a city field with `autocomplete="address-line1"` still gets classified correctly via the HTML `<label>` text). Previously, the first match won unconditionally, so misattributed fields were either mapped to the wrong type or left unfilled.

Fill logic:
  - Added dropdown/select (`<select>`) support — uses `AutofillValue.forList(index)` with three-pass option matching (exact, prefix, reverse prefix) instead of only `forText()`. Previously, filling a dropdown with `forText()` caused an `IllegalStateException` or silently failed, leaving country and state fields empty.
  - Added country code to display name conversion ("US" → "United States") so country dropdowns are filled correctly. Previously, the raw country code did not match any dropdown option.
  - Phone numbers are stripped of whitespace before filling. Previously, stored numbers like "+1 234 567 8901" could fail to fill into fields expecting no spaces.

Save logic:
  - `SaveInfo` uses a single required field (`preferring name`, `email`, or `address`) with the rest as optional. Previously, all detected fields were passed as required, so the `"Save to Brave?"` prompt never appeared unless every single field was filled.
  - `getFieldValue()` reads both text inputs and dropdown selections, so country and state values are saved correctly from `<select>` elements. Previously, only `AUTOFILL_TYPE_TEXT` was handled, so dropdown values were lost and saved as empty strings.

Resolves: https://github.com/brave/brave-browser/issues/41334

<!-- CI-related labels that can be applied to this PR:
* CI/disable-pipeline-step-cache - instruct CI to not cache build steps between runs for the same commit hash
* CI/enable-coverage - enable coverage reporting for your code changes
* CI/enable-test-only-affected - instruct CI to only run tests affected by your change
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-perf-smoke-tests - run smoke performance tests
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/run-teamcity - run TeamCity
* CI/skip-teamcity - skip TeamCity
* CI/skip - do not run CI builds (except noplatform)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

<!--
## Checklist:

- Review design docs
  [Browser design principles](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/chrome_browser_design_principles.md)
  [Style guide](https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++.md)
  [Core principles](https://www.chromium.org/developers/core-principles/)
- Ensure there are (tests)[https://www.chromium.org/developers/testing/]. Unit test as much as possible (including edge cases), but also include browser tests covering high level functionality.
- Ensure that there are comments explaining what classes/methods are/do. The "why" is often more important than the "what" in comments. Also update any relevant docs (moving docs from wiki to brave-core if necessary).
- Request security or other review (third-party libraries, rust code, etc...) if applicable [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) [other review](https://github.com/brave/reviews/issues/new/choose)
  Also see [adding third-party libraries](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/adding_to_third_party.md) for general guidelines on using third party code
- Make sure there is a [ticket](https://github.com/brave/brave-browser/issues) for your issue
- Use Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- Write a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- Squash any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- Add appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- Run `git rebase master` (if needed)
-->

Steps to Reproduce

Fill: Autofill saved addresses into third-party apps
                                                                                                                                                                                                                                                                                                                                                      
Prerequisites:                                                                                                                                                                                                                                                                                                                                      
  - Have at least one address saved in Brave (Settings > Addresses and more > Add address). Include all fields: name, email, phone, street address, city, state, zip, country.
  - Go to Android Settings > Passwords & Autofill > Autofill service > select Brave.
  - In Chrome, enable autofill with third-party apps: Settings > Autofill services > Autofill using another service.

  1. Open Chrome (or any other browser/app)
  2. Navigate to https://fill.dev/form/address
  3. Tap on any address field (e.g. First Name)
  4. Expected: Brave autofill popup appears with saved address profile(s)
  5. Tap on a suggestion
  6. Expected: All matching fields are filled, including text inputs and dropdowns (country, state)

Repeat with https://rsolomakhin.github.io/autofill/ (click "Address" link):
  1. Tap on any field
  2. Expected: Brave autofill popup appears and fills all fields when selected

Save: Save manually entered addresses from third-party apps

Prerequisites: Same as above.

  1. Open Chrome (or any other browser/app)
  2. Navigate to https://fill.dev/form/address
  3. Manually type values into the fields: name, email, phone, address, city, select a state, zip, select a country
  4. Tap Submit
  5. Expected: Android shows a "Save to Brave?" prompt
  6. Tap Save
  7. Open Brave > Settings > Addresses and more
  8. Expected: The new address appears with all fields populated correctly, including country and state from dropdowns

Repeat with https://rsolomakhin.github.io/autofill/ (click "Address" link):
  1. Manually fill the fields and submit
  2. Expected: "Save to Brave?" prompt appears, saved address includes all entered values

https://drive.google.com/file/d/1T1NMjiEEG5Qxc1BvhZZYDe4Nlpf7dcyY/view?usp=sharing

### CODE REVIEW

**Review by AlexeyBarabash — APPROVED** [review]
++

File: browser/autofill/android/java_sources.gni [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+# These sources are included directly into chrome_java rather than the
+# brave/browser/autofill/android:java target to avoid a circular dependency
+# (they depend on ChromeBrowserInitializer which is part of chrome_java).
+brave_browser_autofill_service_java_sources = [
+  "//brave/browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillServiceImpl.java",
+  "//brave/browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillViewStructureParser.java",
```
[the remark this answers is not in the document]
Comment (AlexeyBarabash) [reply browser/autofill/android/java_sources.gni:11]: From the comments above:

>  These sources are included directly into chrome_java rather than the
>  brave/browser/autofill/android:java target to avoid a circular dependency
>  (they depend on ChromeBrowserInitializer which is part of chrome_java).

File: browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+     */
+    private void fillAddressAutofill(
+            Map<String, AutofillId> fields,
+            Map<AutofillId, ViewNode> nodesMap,
+            FillCallback callback) {
+        PersonalDataManager personalDataManager =
+                PersonalDataManagerFactory.getForProfile(
+                        // This is a top-level entry point (AutofillService) with no
+                        // Profile passed in from the framework.
+                        ProfileManager.getLastUsedRegularProfile());
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillServiceImpl.java:201]: service is a top-level entry point

File: browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+        String company =
+                getFieldValue(fields, nodesMap, BraveAutofillViewStructureParser.HINT_COMPANY);
+        String country =
+                getFieldValue(fields, nodesMap, BraveAutofillViewStructureParser.HINT_COUNTRY);
+
+        PersonalDataManager personalDataManager =
+                PersonalDataManagerFactory.getForProfile(
+                        // This is a top-level entry point (AutofillService) with no
+                        // Profile passed in from the framework.
+                        ProfileManager.getLastUsedRegularProfile());
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillServiceImpl.java:396]: service is a top-level entry point

File: browser/autofill/android/java_sources.gni [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2026 The Brave Authors. All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+# These sources are included directly into chrome_java rather than the
+# brave/browser/autofill/android:java target to avoid a circular dependency
+# (they depend on ChromeBrowserInitializer which is part of chrome_java).
+brave_browser_autofill_service_java_sources = [
+  "//brave/browser/autofill/android/java/src/org/chromium/chrome/browser/autofill/BraveAutofillServiceImpl.java",
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply browser/autofill/android/java_sources.gni:10]: that's been resolved here https://github.com/brave/brave-core/pull/35567#discussion_r3089229364

**Review by samartnik — APPROVED** [review]
++

**Review by simoarpe — APPROVED** [review]
👍 Changes look good!
Do you think it may worth adding a small set of tests just for the core functionalities?
I'm not sure if it's something that would be easily doable considering it's a high level service.

**Comment by SergeyZhukovsky** [comment]
> 👍 Changes look good! Do you think it may worth adding a small set of tests just for the core functionalities? I'm not sure if it's something that would be easily doable considering it's a high level service.

The `onFillRequest/onSaveRequest` itself would definitely hard to auto test as they high level entry points of the service. I'm going to extract some parser parts in static functions and unit test them.

**Comment by SergeyZhukovsky** [comment]
> > 👍 Changes look good! Do you think it may worth adding a small set of tests just for the core functionalities? I'm not sure if it's something that would be easily doable considering it's a high level service.
> 
> The `onFillRequest/onSaveRequest` itself would definitely hard to auto test as they high level entry points of the service. I'm going to extract some parser parts in static functions and unit test them.

I added tests here https://github.com/brave/brave-core/pull/35567/changes/1271ceb26e2d479cb79d528956be6ec7b762558f

**Review by netzenbot — APPROVED** [review]
